Array chunk warning

mattchurgin opened the pull request:
- [] Tests passed
- [x] Passes `black dask` / `flake8 dask`


Pull request to address issue #6810 . Issue a warning when mapping a function with 'axis' arguments onto a dask array. This alerts users to possible unexpected behavior when applying a function with axis argument along a dask array axis comprised of more than 1 chunk. Unfortunately not all tests are passing (they passed in my env before this commit). @jsignell would you mind taking a look when you have a chance and seeing if my approach to warning makes sense?

jsignell commented on dask/array/core.py:
```diff
@@ -661,6 +661,22 @@ def map_blocks(
+                else:
+                    warnings.warn(
+                        "Mapping function with axis arguments along axis with nchunks>1 may produce unexpected behavior"
+                    )
```
Probably best not to warn in this case. I think we want to err on the side of warning less often rather than too often. Otherwise they get reallly annoying and people turn them off.

jsignell commented on dask/array/core.py:
```diff
@@ -661,6 +661,22 @@ def map_blocks(
 
     original_kwargs = kwargs
 
+    # issue warning when mapping function with axis argument when array chunks > 1
```
I don't think this comment adds much. I'd rather let the code speak for itself

jsignell commented on dask/array/core.py:
```diff
@@ -661,6 +661,22 @@ def map_blocks(
     original_kwargs = kwargs
 
+    # issue warning when mapping function with axis argument when array chunks > 1
+    if has_keyword(func, "axis"):
+        if len(arrs) == 1:
+            n_axis_chunks = [len(x) for x in arrs[0].chunks]
+            if any([n_chunks > 1 for n_chunks in n_axis_chunks]):
+                if "axis" in original_kwargs.keys():
+                    axis_chunks = n_axis_chunks[original_kwargs["axis"]]
+                    if axis_chunks > 1:
+                        warnings.warn(
+                            "Operating along axis with nchunks>1 may produce unexpected behavior"
+                        )
```
I think this can be simplified some. Perhaps something like:

```suggestion
    if has_keyword(func, "axis") and "axis" in original_kwargs:
        axis = original_kwargs["axis"]
        if len(arrs) == 1:
            if arrs[0].chunksize[axis] != arrs[0].shape[axis]:
                warnings.warn(
                    "Operating along axis with nchunks>1 may produce unexpected behavior"
                )
```

jsignell reviewed (commented):
This seems like a nice addition, but you'll need to make sure that the existing tests pass and add a new test to check this new functionality.

mattchurgin commented:
Thanks @jsignell for the feedback. I incorporated it and added a test. 

I saw some tests were failing, and it looks like that is because the default behavior of pytest here is to throw an error on warnings. When I run pytest with -W ignore::UserWarning, all my tests pass. All tests also pass when I comment out all of my additions except for the warning. I'm not sure if it's possible/reasonable to include an ignore::UserWarning in the pytest config. I'm actually also not sure where pytest is configured in the codebase. What do you think?

jsignell commented:
I think you'd want to treat each test separately by adding a decorator like `@pytest.mark.filterwarnings("ignore:Operating along axis with nchunks>1")`. 

But having seen how much this pattern shows up in the test codebase, I am unsure whether we really want this warning or not (sorry!). I think if there were a way to disable it or an existing pattern of enabling more warnings it could be really nice to have this kind of thing. Maybe someone can chime in with a second opinion? @dask/maintenance

mattchurgin commented:
Thanks @jsignell for the feedback. I appreciate it. Happy to continue working on this if people think it's valuable. I see from the tests that it's assumed users have a good understanding of how the chunk structure of their array influences how computations on it will behave. At the same time, not everyone will do that. I agree we don't want to overdo it with warnings. I leave it up to you and the maintainers to decide whether this use case is overdoing it. Thanks!

jsignell commented:
Ok I am going to close this - I think we want to err on the side of only really important warnings. Thank you for opening it though!
